crates/core/tedge/src/cli/mqtt/error.rs

```diff
@@ -14,4 +16,57 @@ pub enum MqttError {
 
     #[error("MQTT connection error: {0}\n\nHint: Is MQTT server running?")]
     ServerConnection(String),
+
+    #[error("Failed to create a TLS config")]
+    TlsConfig(#[from] CertificateError),
+
+    #[error("Server TLS certificate could not be verified")]
+    ServerCertVerify(#[from] ServerCertVerifyError),
+}
+
+impl From<rumqttc::TlsError> for MqttError {
+    fn from(error: rumqttc::TlsError) -> Self {
+        Self::ServerCertVerify(ServerCertVerifyError::from(error))
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ServerCertVerifyError {
+    /// Received certificate is not in X509v3 format
+    #[error("Received certificate is not in X.509 v3 format")]
+    UnsupportedCertVersion,
+
+    /// Received certificate is not valid for the name it is being validated for.
+    #[error(
+        "Received certificate is not valid for the name it is being validated for.
+        \nHint: Be sure to use a hostname specified in the CN field of the certificate and for the certificate to have a Subject Alt Name section."
+    )]
+    CertNotValidForName,
+
+    #[error(transparent)]
+    Other(rumqttc::TlsError),
+}
+
+// Not sure if there is a better way to do this, unfortunately rumqttc error
+// reporting is a bit awkward
+impl From<rumqttc::TlsError> for ServerCertVerifyError {
```

Comment:
Unfortunately, TLS error reporting is tedious in many crates. See https://github.com/thin-edge/thin-edge.io/blob/8b4755e77cc9dd4aa499784a4d9a88e772d9f2f2/crates/core/tedge/src/cli/certificate/error.rs#L124

Reply:
Yes, the biggest issue is that if we connect via TLS, then all the errors come wrapped inside `rumqttc::ConnectionError::Tls(..)` even if the issue was not with TLS specifically. If it was a TLS error, there is another layer of `rumqttc::TlsError::Io(..)` despite the issue not being related to IO at all, and only then inside we have an actual error which is stringified.

This nesting results in the structure of the error just being the copy of `rumqttc` call stack. All the errors are wrapped in `TlsError(..)` becuase the errors are returned from a function like `TlsClient::read()`, which just wraps all the errors, and that's all the way down the stack.

This is I think an antipattern that I want to look into making sure we're not replicating in thin-edge after I complete this issue. Sure we're the only consumers of our APIs, so the impact is limited, but it's still there, and might cause bad error formatting for the users, if one's not careful when reporting errors.

Reply:
> This is I think an antipattern that I want to look into making sure we're not replicating in thin-edge after I complete this issue. Sure we're the only consumers of our APIs, so the impact is limited, but it's still there, and might cause bad error formatting for the users, if one's not careful when reporting errors.

I would be really happy to have you working on such improvements.